python/ray/data/_internal/iter_batches.py

```diff
@@ -0,0 +1,934 @@
+            the block will never be accessed again.
+        batch_size: Record batch size, or None to let the system pick.
+        batch_format: The format in which to return each batch.
+            Specify "default" to use the current block format (promoting
+            Arrow to pandas automatically), "pandas" to
+            select ``pandas.DataFrame`` or "pyarrow" to select
+            ``pyarrow.Table``. Default is "default".
+        drop_last: Whether to drop the last batch if it's incomplete.
+        collate_fn: A function to apply to each data batch before returning it.
+        shuffle_buffer_min_size: If non-None, the data will be randomly shuffled using a
+            local in-memory shuffle buffer, and this value will serve as the minimum
+            number of rows that must be in the local in-memory shuffle buffer in order
+            to yield a batch.
+        shuffle_seed: The seed to use for the local random shuffle.
+        ensure_copy: Whether batches are always copied from the underlying base
+            blocks (not zero-copy views).
+        prefetch_batches: The number of batches to fetch ahead of the current batch to
+            process. If set to greater than 0, a separate thread will be used to fetch
+            the specified amount of formatted batches from blocks. This improves
+            performance for non-CPU bound UDFs, allowing batch fetching compute and
+            formatting to be overlapped with the UDF. Defaults to 0 (no prefetching
+            enabled).
+
+    Returns:
+        An iterator over record batches.
+    """
+    context = DatasetContext.get_current()
+
+    if (
+        prefetch_batches > 0
+        and context.actor_prefetcher_enabled
+        and not ray.util.client.ray.is_connected()
+    ):
+        prefetcher = ActorBlockPrefetcher()
+    else:
+        prefetcher = WaitBlockPrefetcher()
+
+    def _async_iter_batches(block_refs):
+        # Step 1: Construct logical batches based on the metadata.
+        batch_iter = _bundle_block_refs_to_logical_batches(
+            block_refs, batch_size=batch_size, drop_last=drop_last
+        )
+
+        # Step 2: Shuffle the logical batches if applicable.
+        if shuffle_buffer_min_size is not None:
+            batch_iter = _local_shuffle_logical_batches(
+                shuffle_buffer_min_size=shuffle_buffer_min_size,
+                shuffle_seed=shuffle_seed,
+            )
+
+        # Step 3: Prefetch logical batches locally.
+        if prefetch_batches > 0:
+            batch_iter = _prefetch_batches_locally(
+                batch_iter,
+                prefetcher=prefetcher,
+                num_batches_to_prefetch=prefetch_batches,
+                stats=stats,
+            )
+
+        def threadpool_computations(logical_batch_iter: Iterator[LogicalBatch]):
+            # Step 4.1: Resolve the blocks.
+            resolved_batch_iter = _resolve_blocks(
```

Comment:
It seems you do block deletion here, but isn't there a risk of a race condition if a block gets deleted that another thread still needs? I would expect you would have to do deletion at the end in serial order / from a single thread.